pymeasure/instruments/common_base.py

```diff
@@ -0,0 +1,553 @@
+    :param fset_params_list: List of parameter names that are dynamically configurable
+    :param prefix: String to be prefixed to get dynamically configurable
+                   parameters.
+    """
+
+    def __init__(self, fget=None, fset=None, fdel=None, doc=None, fget_params_list=None,
+                 fset_params_list=None, prefix=""):
+        super().__init__(fget, fset, fdel, doc)
+        self.fget_params_list = () if fget_params_list is None else fget_params_list
+        self.fset_params_list = () if fset_params_list is None else fset_params_list
+        self.name = ""
+        self.prefix = prefix
+
+    def __get__(self, obj, objtype=None):
+        if obj is None:
+            # Property return itself when invoked from a class
+            return self
+        if self.fget is None:
+            raise AttributeError(f"Unreadable attribute {self.name}")
+
+        kwargs = {}
+        for attr in self.fget_params_list:
+            attr_instance_name = self.prefix + "_".join([self.name, attr])
+            if hasattr(obj, attr_instance_name):
+                kwargs[attr] = getattr(obj, attr_instance_name)
+        return self.fget(obj, **kwargs)
+
+    def __set__(self, obj, value):
+        if self.fset is None:
+            raise AttributeError(f"Can't set attribute {self.name}")
+        kwargs = {}
+        for attr in self.fset_params_list:
+            attr_instance_name = self.prefix + "_".join([self.name, attr])
+            if hasattr(obj, attr_instance_name):
+                kwargs[attr] = getattr(obj, attr_instance_name)
+        self.fset(obj, value, **kwargs)
+
+    def __set_name__(self, owner, name):
+        self.name = name
+
+
+class CommonBase:
+    """Base class for instruments and channels.
+
+    This class contains everything needed for pymeasure's property creator
+    :meth:`control` and its derivatives :meth:`measurement` and :meth:`setting`.
+
+    :param setup: A method executed between instantiation and creation of children (channels).
+    :param setup_kwargs: Keyword arguments for the setup method.
+    """
+
+    # Variable holding the list of DynamicProperty parameters that are configurable
+    # by users
+    _fget_params_list = ('get_command',
+                         'values',
+                         'map_values',
+                         'get_process',
+                         'command_process',
+                         'check_get_errors')
+
+    _fset_params_list = ('set_command',
+                         'validator',
+                         'values',
+                         'map_values',
+                         'set_process',
+                         'command_process',
+                         'check_set_errors')
+
+    # Prefix used to store reserved variables
+    __reserved_prefix = "___"
+
+    def __init__(self, setup=None, setup_kwargs={}):
+        self._special_names = self._setup_special_names()
+        if setup is not None:
+            setup(**setup_kwargs)
```

Comment:
I think `self` should be passed to `setup` in order to allow command execution, right ?
Is it possible to remove completely `setup_kwargs` and just pass `self` ?

Reply:
I thought, that you pass the instance method, e.g.
~~~
super().__init__(self, setup=self.doing_setup_stuff, setup_kwargs={'something': 5})
~~~
That way, you already hand over self.
I do not know, whether the kwargs are necessary, but I wanted to include the possibility (maybe to depend on some instrument argument etc.)

Reply:
Make sense to me. Thanks